### PR TITLE
#3271 nominee link bank account

### PR DESCRIPTION
**Pull request by AlexeyTuboltsev**
This PR
- adds the linkBankAccount step to nominee flow
- refactors the mobile/authorized menus to include the nominee user type

### CODE REVIEW

File: app/components/nominee-dashboard/NomineeDashboardContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+interface ILinkedNomineeStateProps {
+  eto: TEtoWithCompanyAndContract | undefined;
+}
+
+interface ILinkedNomineeComponentProps {
+  eto: TEtoWithCompanyAndContract;
+}
+
+export const AccountSetupContainer: React.FunctionComponent = ({ children }) => (
```
Comment (desfero) [inline app/components/nominee-dashboard/NomineeDashboardContainer.tsx]: duplicate of `AccountSetupContainer.tsx`

File: app/components/nominee-dashboard/NomineeDashboardContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  <div data-test-id="nominee-dashboard" className={styles.accountSetupWrapper}>
+    {children}
+  </div>
+);
+
+const NotLinkedNomineeDashboardContainer: React.FunctionComponent = ({ children }) => (
```
Comment (desfero) [inline app/components/nominee-dashboard/NomineeDashboardContainer.tsx]: duplicate of `NotLinkedNomineeDashboardContainer.tsx`

File: app/components/nominee-dashboard/NomineeDashboardContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  <div data-test-id="nominee-dashboard" className={styles.nomineeDashboardContainer}>
+    {children}
+  </div>
+);
+
+export const LinkedNomineeContentPanel: React.FunctionComponent = ({ children }) => (
```
Comment (desfero) [inline app/components/nominee-dashboard/NomineeDashboardContainer.tsx]: duplicate of `LinkedNomineeContentPanel.tsx`

File: app/components/nominee-dashboard/NomineeDashboardContainer.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        dispatch(actions.eto.getNomineeEtos());
+      },
+    }),
+  )(WrappedComponent);
+
+const LinkedNomineeDashboardContainer = connectToNomineeEto<{}>(
+  LinkedNomineeDashboardContainerLayout,
+);
+
+const NomineeDashboardContainer: React.FunctionComponent<IExternalProps> = ({
```
Comment (desfero) [inline app/components/nominee-dashboard/NomineeDashboardContainer.tsx]: not used?

File: app/components/nominee-dashboard/NomineeDashboardContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    </h2>
+    <h2 className={styles.columnTitle}>
+      <FormattedMessage id="nominee-flow.linked-nominee-layout.issuer-campaign" />
+    </h2>
+    <LinkedNomineeContentPanel>{children}</LinkedNomineeContentPanel>
+    {eto && <NomineeEtoOverviewThumbnail eto={eto} shouldOpenInNewWindow={false} />}
+  </div>
+);
+
+const connectToNomineeEto = <T extends {}>(
```
Comment (desfero) [inline app/components/nominee-dashboard/NomineeDashboardContainer.tsx]: this is duplicated

File: app/components/layouts/menus/mobileMenu/MobileMenuInternal.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import * as React from "react";
+import { compose, withProps } from "recompose";
+
+import { DropdownMenuEntry } from "../MenuEntry";
+import { initMobileMenuData } from "../utils";
+import { IMenuData, IMenuDataByUserType, IMobileMenuExternalProps } from "./MobileMenu";
+
+import * as styles from "./MobileMenu.module.scss";
+
+export const MenuLayout: React.FunctionComponent<IMenuData> = ({ data }) => (
```
Comment (desfero) [inline app/components/layouts/menus/mobileMenu/MobileMenuInternal.tsx]: `IMenuData` can be moved here from `MobileMenu`

File: app/components/AppRouter.tsx [hunk 7]
```diff
@@ -142,9 +143,20 @@ export const AppRouter: React.FunctionComponent = () => (
...
     {/* only issuer routes */}
-    <OnlyAuthorizedRoute path={appRoutes.documents} issuerComponent={Documents} />
     <OnlyAuthorizedRoute path={appRoutes.etoRegister} issuerComponent={EtoRegister} />
-    <OnlyAuthorizedRoute path={appRoutes.etoIssuerView} issuerComponent={EtoIssuerView} exact />
+
+    {/* issuer AND nominee routes */}
+    <OnlyAuthorizedRoute
+      path={appRoutes.documents}
+      issuerComponent={Documents}
+      nomineeComponent={Documents}
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/AppRouter.tsx]: yes,
done

File: app/components/eto/EtoNomineeView.tsx [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+type TStateProps = {
+  eto: TEtoWithCompanyAndContract | undefined;
+};
+
+type TViewProps = {
+  eto: TEtoWithCompanyAndContract;
+  publicView: boolean;
+};
+
+interface ILinkedNomineeComponentProps {
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/EtoNomineeView.tsx]: done

File: app/components/eto/EtoNomineeView.tsx [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+export const connectToNomineeEto = <T extends {}>(
+  WrappedComponent: React.ComponentType<ILinkedNomineeComponentProps & T>,
+) =>
+  compose<ILinkedNomineeComponentProps & T, T>(
+    createErrorBoundary(ErrorBoundaryLayout),
+    appConnect<ILinkedNomineeStateProps, {}, T>({
+      stateToProps: state => {
+        const nomineeId = selectUserId(state);
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/EtoNomineeView.tsx]: done

File: app/components/eto/EtoNomineeView.tsx [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        if (nomineeId) {
+          return {
+            eto: selectEtoOfNominee(state, nomineeId),
+          };
+        } else {
+          throw new Error("linked nominee eto id is invalid");
+        }
+      },
+    }),
+    onEnterAction<{}>({
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/EtoNomineeView.tsx]: fixed

File: app/components/eto/EtoNomineeView.tsx [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+      actionCreator: dispatch => {
+        dispatch(actions.eto.getNomineeEtos());
+      },
+    }),
+  )(WrappedComponent);
+
+export const EtoNomineeViewWithContainer = compose<TViewProps, ILinkedNomineeComponentProps>(
+  withProps<{ publicView: boolean }, TStateProps>(() => ({ publicView: false })),
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/EtoNomineeView.tsx]: done

File: app/components/eto/EtoNomineeView.tsx [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+      },
+    }),
+    onEnterAction<{}>({
+      actionCreator: dispatch => {
+        dispatch(actions.eto.getNomineeEtos());
+      },
+    }),
+  )(WrappedComponent);
+
+export const EtoNomineeViewWithContainer = compose<TViewProps, ILinkedNomineeComponentProps>(
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/EtoNomineeView.tsx]: done, if I got you right

File: app/components/layouts/header/Header.tsx [hunk 13]
```diff
@@ -8,8 +8,8 @@ import { appConnect } from "../../../store";
 import { appRoutes } from "../../appRoutes";
 import { ButtonLink, EButtonTheme } from "../../shared/buttons/index";
 import { loginWalletRoutes, walletRegisterRoutes } from "../../wallet-selector/walletRoutes";
-import { MenuAuthorized } from "../menus/MenuAuthorized";
-import { MobileMenu } from "../menus/MobileMenu";
+import { MenuAuthorized } from "../menus/menuAuthorized/MenuAuthorized";
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/layouts/header/Header.tsx]: just renamed it to menu

File: app/components/layouts/menus/menuAuthorized/MenuAuthorized.tsx [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+interface IStateProps {
+  userType: EUserType;
+  actionRequired: boolean;
+}
+
+interface IWithProps {
+  isLinkActive: (match: match<unknown>) => boolean;
+}
+
+const LayoutAuthorizedMenuComponent: React.FunctionComponent<IStateProps & IWithProps> = ({
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/layouts/menus/menuAuthorized/MenuAuthorized.tsx]: I renamed them all

File: app/components/layouts/menus/menuAuthorized/MenuAuthorized.tsx [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        throw new Error("user type is invalid");
+      }
+    },
+  }),
+  withHandlers<IStateProps, IWithProps>({
+    isLinkActive: () => match => Boolean(match),
+  }),
+)(LayoutAuthorizedMenuComponent);
+
+export { InvestorMenuAuthorized, MenuAuthorized, LayoutAuthorizedMenuComponent };
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/layouts/menus/menuAuthorized/MenuAuthorized.tsx]: sorry, it's a leftover

File: app/components/layouts/menus/mobileMenu/MobileMenu.tsx [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+  />
+);
+
+const MobileMenu = compose<IUserType & IMobileMenuExternalProps, {}>(
+  appConnect<IStateProps, IDispatchProps>({
+    stateToProps: state => {
+      const userType = selectUserType(state);
+      if (userType !== undefined) {
+        return {
+          userType: selectUserType(state),
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/layouts/menus/mobileMenu/MobileMenu.tsx]: done

File: app/components/layouts/menus/mobileMenu/MobileMenu.tsx [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+      }
+    },
+    dispatchToProps: dispatch => ({
+      logout: () => {
+        dispatch(actions.auth.logout());
+      },
+    }),
+  }),
+  withHandlers<IStateProps & IDispatchProps, IHandlers>({
+    isLinkActive: () => match => Boolean(match),
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/layouts/menus/mobileMenu/MobileMenu.tsx]: done

File: app/components/layouts/menus/mobileMenu/MobileMenuInternal.tsx (hunk 6 above)
Comment (AlexeyTuboltsev) [reply app/components/layouts/menus/mobileMenu/MobileMenuInternal.tsx]: done

File: app/components/nominee-dashboard/NomineeDashboardContainer.tsx [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+import { actions } from "../../modules/actions";
+import { selectUserId } from "../../modules/auth/selectors";
+import { selectEtoOfNominee } from "../../modules/eto/selectors";
+import { TEtoWithCompanyAndContract } from "../../modules/eto/types";
+import { appConnect } from "../../store";
+import { onEnterAction } from "../../utils/OnEnterAction";
+import { NomineeEtoOverviewThumbnail } from "../eto/overview/EtoOverviewThumbnail/EtoOverviewThumbnail";
+import { ENomineeTask } from "./NomineeTasksData";
+
+import * as styles from "./NomineeDashboard.module.scss";
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/nominee-dashboard/NomineeDashboardContainer.tsx]: this whole file is a duplicate, I'm sorry

File: app/components/nominee-dashboard/NomineeDashboardContainer.tsx (hunk 1 above)
Comment (AlexeyTuboltsev) [reply app/components/nominee-dashboard/NomineeDashboardContainer.tsx]: this whole file is a duplicate, I'm sorry

File: app/components/nominee-dashboard/NomineeDashboardContainer.tsx (hunk 2 above)
Comment (AlexeyTuboltsev) [reply app/components/nominee-dashboard/NomineeDashboardContainer.tsx]: this whole file is a duplicate, I'm sorry

File: app/components/nominee-dashboard/NomineeDashboardContainer.tsx (hunk 3 above)
Comment (AlexeyTuboltsev) [reply app/components/nominee-dashboard/NomineeDashboardContainer.tsx]: this whole file is a duplicate, I'm sorry

File: app/components/nominee-dashboard/NomineeDashboardContainer.tsx [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+
+export const LinkedNomineeContentPanel: React.FunctionComponent = ({ children }) => (
+  <section className={styles.dashboardContentPanel}>{children}</section>
+);
+
+const LinkedNomineeDashboardContainerLayout: React.FunctionComponent<
+  ILinkedNomineeComponentProps
+> = ({ children, eto }) => (
+  <div data-test-id="nominee-dashboard" className={styles.linkedNomineeDashboardContainer}>
+    <h2 className={styles.columnTitle}>
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/nominee-dashboard/NomineeDashboardContainer.tsx]: this whole file is a duplicate, I'm sorry

File: app/components/nominee-dashboard/NomineeDashboardContainer.tsx (hunk 5 above)
Comment (AlexeyTuboltsev) [reply app/components/nominee-dashboard/NomineeDashboardContainer.tsx]: this whole file is a duplicate, I'm sorry

File: app/components/nominee-dashboard/NomineeDashboardContainer.tsx (hunk 4 above)
Comment (AlexeyTuboltsev) [reply app/components/nominee-dashboard/NomineeDashboardContainer.tsx]: this whole file is a duplicate, I'm sorry

File: app/components/nominee-dashboard/linkBankAccount/LinkBankAccount.tsx [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+
+import * as link from "../../../assets/img/inline_icons/social_link.svg";
+
+interface IProps {
+  verifyBankAccount: () => void;
+}
+
+interface IExternalProps {
+  verifyBankAccount: () => void;
+  isBankAccountVerified: boolean;
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/nominee-dashboard/linkBankAccount/LinkBankAccount.tsx]: they are used on the line 44

File: app/components/nominee-dashboard/linkToIssuer/utils.ts [hunk 21]
```diff
@@ -68,20 +68,20 @@ export const getNomineeRequestComponentState = (
     nomineeRequest.state === ENomineeRequestStatus.APPROVED &&
     !nomineeEto
   ) {
-    return ENomineeRequestComponentState.CREATE_NEW_REQUEST;
+    return ENomineeRequestComponentState.CREATE_ANOTHER_REQUEST;
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/nominee-dashboard/linkToIssuer/utils.ts]: all of the business logic will be moved to sagas soon, i suggest I'll leave it here for now.
Here's the ticket: #3286

File: app/modules/auth/jwt/sagas.ts [hunk 22]
```diff
@@ -146,6 +146,7 @@ export function* ensurePermissionsArePresentAndRunEffect(
   title: TMessage,
   message: TMessage,
   inputLabel?: TMessage,
+  recoverAction?: StringableActionCreator<TAction>,
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/modules/auth/jwt/sagas.ts]: a leftover, sorry

File: app/modules/eto-flow/actions.ts [hunk 23]
```diff
@@ -7,7 +7,10 @@ import { TEtoProducts } from "../../lib/api/eto/EtoProductsApi.interfaces";
 import { createActionFactory } from "../actionsUtils";
 
 export const etoFlowActions = {
-  loadIssuerEto: createActionFactory("ETO_FLOW_LOAD_ISSUER_ETO"),
+  loadIssuerEto: createActionFactory(
+    "ETO_FLOW_LOAD_ISSUER_ETO",
+    (inBackground: boolean = false) => ({ inBackground }),
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/modules/eto-flow/actions.ts]: fixed

File: app/modules/eto/selectors.ts [hunk 24]
```diff
@@ -187,7 +187,7 @@ export const selectEtoOfNominee = (
 ): TEtoWithCompanyAndContract | undefined => {
   const etoState = selectEtoState(state);
   const previewCode = Object.keys(etoState.etos).find(
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/modules/eto/selectors.ts]: done

File: app/utils/renderComponentWithProps.unsafe.tsx [hunk 25]
```diff
@@ -0,0 +1,9 @@
+import * as React from "react";
+import { ComponentEnhancer, renderComponent, withProps } from "recompose";
+
+export function renderComponentWithPropsUnsafe<O, I>(
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/utils/renderComponentWithProps.unsafe.tsx]: I'll use it later, I promise 😁

**Pull request merged**